File: guests/haskell/richrows.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# OPTIONS_GHC -Wno-missing-signatures #-}
-- A key path's wire tag (KayaWire.Value) has no spelling a guest may
-- write (tools/check-sugar-surface.py's wire-tag clause) — the
-- fromWire-decoding helper below is left unsigned so its argument
-- type is inferred, never named.

-- The rich rows scene, Haskell port — guests/rust/richrows.rs,
-- tools/scenes/richrows.steps: a rich textarea per stamped ROW whose
-- document is a FIELD of the row (docs/rich-text-plan.md §19). The app
-- writes a copy's document by patching its row, and a copy's own act
-- folds into the row the app reads back.

import GHC.Generics (Generic)

import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

data Note = Note {title :: Text, body :: Document}
  deriving stock (Generic)
  deriving anyclass (KayaRecord)


-- The core's spelling of runs (`expect_runs`), so the row's field and the
-- core's mirror are compared as one string.
spell :: [Run] -> Text
spell = T.intercalate "|" . map one
  where
    one r
      | runValue r == "true" = span_ r <> runName r
      | otherwise = span_ r <> runName r <> "=" <> runValue r
    span_ r = T.pack (show (runStart r) ++ ":" ++ show (runEnd r) ++ " ")

-- The key path's own wire tag decoded through 'fromWire', never named:
-- the argument's type (the wire's Value) has no spelling a guest may
-- write (tools/check-sugar-surface.py's wire-tag clause) — inferred here
-- from its call sites instead.
keyText v = fromWire v :: Text

noteAt items key = case lookup key items of
  Just note -> note
  Nothing -> error ("richrows: no row " ++ T.unpack (keyText key))

main :: IO ()
main = kayaMain $ \app -> do
  (notes, lastAct, bodyNode) <- buildTx app $ do
    notes <- collectionOf @Note
    lastAct <- signal ("" :: Text)
    view <- signal ("" :: Text)

    -- An undo or redo moved the row back: the app reads ITS OWN mirror of
    -- row b, which is the fold a restored Blob field lands in.
    let restored _label _delta = submitTx app $ do
          items <- recordItems notes
          let note = noteAt items (toWire ("b" :: Text))
          writeSignal
            view
            (docText (body note) <> " | " <> spell (docRuns (body note)))
    window
      primary
      [ WTitle "richrows",
        WMenus
          [ menu
              "Edit"
              []
              [ item "Undo" [IRole roleUndo],
                item "Redo" [IRole roleRedo]
              ]
          ],
        WOnUndone restored,
        WOnRedone restored
      ]

    lastLabel <- labelBound lastAct -- label#0
    viewLabel <- labelBound view -- label#1
    buttons <-
      row
        []
        [ -- button#0 — the app writes a copy by patching its row
          buttonOn "patch b" . undoableTx app "patch b" $
            patch
              notes
              ("b" :: Text)
              [set (field @"body" @Note) (italicRun (0, 7) (documentOf "Patched"))],
          -- button#1 — the row the copy's own act folded into
          buttonOn "read a" . submitTx app $ do
            items <- recordItems notes
            let note = noteAt items (toWire ("a" :: Text))
            writeSignal
              view
              (docText (body note) <> " | " <> spell (docRuns (body note)))
        ]
    (rows, bodyNode) <- forEach (recordHandle notes) $ do
      titleLabel <- label (field @"title" @Note)
      bodyNode <- withTplAttrs [TplA11yId ("body" :: Text)] (textareaRichBound (field @"body" @Note))
      _ <- columnOf [pure titleLabel, pure bodyNode]
      return bodyNode

    root <- column [] [pure lastLabel, pure viewLabel, pure buttons, pure rows]
    mount root

    insertRecord notes ("a" :: Text) (Note "a" (boldRun (0, 6) (documentOf "Héllo world")))
    insertRecord
      notes
      ("b" :: Text)
      (Note "b" (linkRun (7, 11) "https://kaya.dev" (documentOf "Second note")))
    return (notes, lastAct, bodyNode)

  -- The row's field already carries the copy's act when these fire: the
  -- app reads the row, never the widget.
  let acted [] = error "kaya: acted's key path is never empty"
      acted (key : _) = submitTx app $ do
        items <- recordItems notes
        let note = noteAt items key
        writeSignal
          lastAct
          (keyText key <> ": " <> spell (docRuns (body note)))
  onEditNode app bodyNode (\keys _ -> acted keys)
  onFormatNode app bodyNode (\keys _ -> acted keys)

File: guests/rust/richrows.rs
//! The rich rows scene (tools/scenes/richrows.steps): a rich textarea per
//! stamped ROW whose document is a FIELD of the row (docs/rich-text-plan.md
//! §19). The app writes a copy's document by patching its row, and a copy's
//! own act folds into the row the app reads back.

use kaya::PathKey;

#[derive(kaya::KayaGen, Clone, Debug, PartialEq)]
struct Note {
    title: String,
    body: kaya::Document,
}

#[derive(Clone)]
enum Msg {
    Acted(kaya::Path),
    Patch,
    Read,
    /// An undo or redo moved the row back: the app reads ITS OWN mirror of
    /// row b, which is the fold a restored Blob field lands in.
    Restored,
}

/// The core's spelling of runs (`expect_runs`), so the row's field and the
/// core's mirror are compared as one string.
fn spell(runs: &[kaya::Run]) -> String {
    runs.iter()
        .map(|run| {
            if run.is_flag() {
                format!("{}:{} {}", run.range.start, run.range.end, run.name)
            } else {
                format!("{}:{} {}={}", run.range.start, run.range.end, run.name, run.value)
            }
        })
        .collect::<Vec<_>>()
        .join("|")
}

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::new();
    let (notes, last, view) = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW)
            .title("richrows")
            .menu("Edit", |m| {
                m.item("Undo").role(kaya::MenuRole::Undo).id();
                m.item("Redo").role(kaya::MenuRole::Redo).id();
            })
            .id();
        let notes = tx.collection::<Note>();
        let last = tx.signal("");
        let view = tx.signal("");
        let root = tx
            .column(|tx| {
                tx.label(last);
                tx.label(view);
                tx.row(|tx| {
                    let patch = tx.button("patch b").id();
                    msgs.on_click(patch, Msg::Patch);
                    let read = tx.button("read a").id();
                    msgs.on_click(read, Msg::Read);
                });
                for mut row in notes.rows(tx) {
                    row.column(|t| {
                        t.label(Note::title());
                        let body = t.textarea_rich_bound(Note::body());
                        t.a11y_id(body, "body");
                        msgs.on_edit_node(body, |path, _| Msg::Acted(path));
                        msgs.on_format_node(body, |path, _| Msg::Acted(path));
                    });
                }
            })
            .id();
        tx.mount(root);
        tx.insert(
            &notes,
            "a",
            Note {
                title: "a".to_owned(),
                body: kaya::Document::new("Héllo world").mark(0..6, "bold", true),
            },
        );
        tx.insert(
            &notes,
            "b",
            Note {
                title: "b".to_owned(),
                body: kaya::Document::new("Second note").mark(7..11, "link", "https://kaya.dev"),
            },
        );
        (notes, last, view)
    });

    msgs.on_undone(kaya::DEFAULT_WINDOW, |_, _| Msg::Restored);
    msgs.on_redone(kaya::DEFAULT_WINDOW, |_, _| Msg::Restored);

    let row = |tx: &kaya::Tx<'_>, key: &str| -> Note {
        notes
            .get(tx, key)
            .unwrap_or_else(|| panic!("richrows: no row {key:?}"))
    };
    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            // The row's field already carries the copy's act when this
            // fires: the app reads the row, never the widget.
            Msg::Acted(path) => ctx.apply(|tx| {
                let key = path.key::<String>(0);
                let note = row(tx, &key);
                let shown = format!("{key}: {}", spell(&note.body.runs));
                tx.write(last, shown);
            }),
            Msg::Patch => ctx.apply(|tx| {
                tx.undoable("patch b");
                notes.patch(tx, "b").body(kaya::Document::new("Patched").mark(0..7, "italic", true));
            }),
            Msg::Read => ctx.apply(|tx| {
                let note = row(tx, "a");
                let shown = format!("{} | {}", note.body.text, spell(&note.body.runs));
                tx.write(view, shown);
            }),
            Msg::Restored => ctx.apply(|tx| {
                let note = row(tx, "b");
                let shown = format!("{} | {}", note.body.text, spell(&note.body.runs));
                tx.write(view, shown);
            }),
        }
    }
}

fn main() {
    kaya::run(app)
}
